### PR TITLE
frontend: Integrate Safe

### PR DESCRIPTION
**Pull request by manuelwedler**
see #1507
(we still need to open an issue against their repo to get listed in the Safe App)

This was much more work than expected. Some function declarations and interfaces of our internal architecture needed to get changed. A few things are still missing: ~~The warning if the target address is the same as the Safe's one is missing. Some tests are missing.~~ Maybe it would be nicer to use a flag on the `IEthereumProvider` interface if the disconnect should be shown or if the target address should be prefilled.

### CODE REVIEW

**Comment by manuelwedler** [comment]
Added a commit for showing the warning if the same address as the connected Safe is chosen for the target address.

**Review by GabrielBuragev — CHANGES_REQUESTED** [review]
Nice! Some amazing improvements in the structure of the providers and i can already see its paying off. Great job man.

I have a decent amount of nitpicky suggestions and some questions. Once they are addressed happily approving!

File: frontend/src/services/web3-provider/safe-provider.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
+import type { SafeInfo } from '@safe-global/safe-apps-sdk';
+
+import { BasicEthereumProvider } from '@/services/web3-provider';
+import { SafeAppProvider, SafeAppsSDK } from '@/services/web3-provider/util-export';
+
+export async function createSafeProvider(): Promise<SafeProvider | undefined> {
+  const sdk = new SafeAppsSDK();
+  // check if we're in an iframe
+  if (window?.parent === window) {
+    console.error('shit');
```
[the remark this answers is not in the document]
Comment (manuelwedler) [reply frontend/src/services/web3-provider/safe-provider.ts:10]: lol whoops

File: frontend/src/components/RequestSourceInputs.vue [hunk 2]
```diff
@@ -187,16 +188,23 @@ const selectedSourceChain = computed({
...
+  () => !provider.value || !!provider.value.switchChainSafely,
+);
+
 // Need to switch chain in case the user selected one before connecting a wallet
-watch(providerChainOption, (_newProviderChainOption, previousProviderChainOption) => {
+watch(providerChainOption, async (_newProviderChainOption, previousProviderChainOption) => {
   if (provider.value && !previousProviderChainOption && _selectedSourceChain.value) {
-    const switchSuccessful = provider.value.switchChainSafely(_selectedSourceChain.value.value);
+    let switchSuccessful = false;
+    if (provider.value.switchChainSafely) {
```
[the remark this answers is not in the document]
Comment (manuelwedler) [reply frontend/src/components/RequestSourceInputs.vue:205]: Do you mean a getter on the store that would check if `switchChainSafely` is available on the provider? So the function would still be optional on the `IEthereumProvider` interface. In this case, I think Typescript would still make the `if (provider.value.switchChainSafely)` check required as the typing cannot be inferred from the store getter.

[the remark this answers is not in the document]
Comment (manuelwedler) [reply frontend/src/components/RequestSourceInputs.vue:205]: The flags I meant should be implemented on the interface and don't relate to the existence of a function on the provider.

File: frontend/src/services/transactions/request-manager.ts [hunk 3]
```diff
@@ -129,14 +134,16 @@ export async function sendRequestTransaction(
...
-
-    await transaction.wait(getConfirmationTimeBlocksForChain(chainId));
+    const transaction = await requestManagerContract.createRequest(...requestParameter, {
+      gasLimit: estimatedGasLimit,
+    });
 
-    return transaction.hash;
+    const transactionReceipt = await transaction.wait(getConfirmationTimeBlocksForChain(chainId));
+    let transactionHash = transactionReceipt.transactionHash;
+    if (provider.getActualTransactionHash) {
```
[the remark this answers is not in the document]
Comment (manuelwedler) [reply frontend/src/services/transactions/request-manager.ts:143]: Yeah I think it is nice to have optional methods on the provider interface that allow to modify behavior in some places.

File: frontend/tests/unit/services/web3-provider/safe-provider.spec.ts [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+      expect(provider).toBeInstanceOf(SafeProvider);
+      expect(provider?.chainId.value).toBe(chain.identifier);
+      expect(provider?.signer.value).toBe(signer);
+      expect(provider?.signerAddress.value).toBe(signerAddress);
+    });
+
+    it('returns undefined when not connected to a safe', async () => {
+      const sdk = new MockedSafeAppsSDK();
+      // eslint-disable-next-line @typescript-eslint/no-empty-function
+      sdk.safe.getInfo = new Promise(function () {});
```
[the remark this answers is not in the document]
Comment (manuelwedler) [reply frontend/tests/unit/services/web3-provider/safe-provider.spec.ts:70]: Cool, thanks!

File: frontend/src/services/transactions/request-manager.ts [hunk 5]
```diff
@@ -99,7 +100,7 @@ export async function getRequestFee(
 }
 
 export async function sendRequestTransaction(
-  signer: JsonRpcSigner,
+  provider: IEthereumProvider,
```
[the remark this answers is not in the document]
Comment (manuelwedler) [reply frontend/src/services/transactions/request-manager.ts:103]: Sure!

File: frontend/src/components/RequestSourceInputs.vue (hunk 2 above)
[the remark this answers is not in the document]
Comment (GabrielBuragev) [reply frontend/src/components/RequestSourceInputs.vue:205]: Ignoring this as discussed. Will not provide much benefit.

**Pull request merged**